src/backend/dx12/src/command.rs

```diff
@@ -1283,61 +1285,82 @@ impl com::RawCommandBuffer<Backend> for CommandBuffer {
-                } else {
-                    region.buffer_width
-                };
+            let r = region.borrow();
 
-                let buffer_height = if region.buffer_height == 0 {
-                    region.image_extent.height
-                } else {
-                    region.buffer_height
-                };
+            assert_eq!(r.buffer_offset % d3d12::D3D12_TEXTURE_DATA_PLACEMENT_ALIGNMENT as u64, 0);
+            assert!(r.image_extent.width < width as u32, r.image_extent.height <= height as u32 && r.image_extent.depth <= depth as u32);
+
+            let buffer_width = if r.buffer_width == 0 {
+                r.image_extent.width
+            } else {
+                r.buffer_width
+            };
+            let buffer_height = if r.buffer_height == 0 {
+                r.image_extent.height
+            } else {
+                r.buffer_height
+            };
 
-                assert!(buffer_width >= width as u32);
-                assert_eq!(region.buffer_offset % d3d12::D3D12_TEXTURE_DATA_PLACEMENT_ALIGNMENT as u64, 0);
+            if aa != image::AaMode::Single ||
+                r.image_layers.aspects.contains(Aspects::DEPTH) ||
+                r.image_layers.aspects.contains(Aspects::STENCIL)
+            {
+                if r.image_offset.x + r.image_offset.y + r.image_offset.z != 0 ||
+                    r.image_extent.width != width as u32 || r.image_extent.height != height as u32 || r.image_extent.depth != depth as u32
+                {
+                    error!("MSAA and depth/stencil copies have to cover the whole subresource");
+                }
+            }
 
-                let row_pitch = div(buffer_width, image.block_dim.0 as _) * image.bytes_per_block as u32;
-                let slice_pitch = div(buffer_height, image.block_dim.1 as _) * row_pitch;
-                assert_eq!(row_pitch % d3d12::D3D12_TEXTURE_DATA_PITCH_ALIGNMENT as u32, 0);
+            assert!(buffer_width >= width as u32);
+            let row_pitch = div(buffer_width, image.block_dim.0 as _) * image.bytes_per_block as u32;
+            let slice_pitch = div(buffer_height, image.block_dim.1 as _) * row_pitch;
+            assert_eq!(row_pitch % d3d12::D3D12_TEXTURE_DATA_PITCH_ALIGNMENT as u32, 0);
 
-                let height = height as _;
-                let depth = depth as _;
+            let width_offset_bytes = div(r.image_offset.x as _, image.block_dim.0 as _) * image.bytes_per_block as u32;
+            assert!(width_offset_bytes <= r.buffer_offset as u32);
 
+            let (footprint_height, height_offset) = if r.image_extent.depth <= 1 {
+                // when a single depth slice is copied, we can pretend to deal with a smaller image
+                (r.image_extent.height, 0)
+            } else {
+                (height as u32, r.image_offset.y as u32)
+            };
+
+            let src_box = d3d12::D3D12_BOX {
+                left: r.image_offset.x as _,
```

Comment:
This looks incorrect, should be 0. image offset is specified as parameter on CopyTextureRegion.